linfa-clustering/src/dbscan/algorithm.rs

```diff
@@ -0,0 +1,116 @@
+use crate::dbscan::hyperparameters::DBScanHyperParams;
+use ndarray::{Array1, ArrayBase, ArrayView, Axis, Data, Ix1, Ix2};
+use ndarray_stats::DeviationExt;
+
+pub fn dbscan(
+    hyperparameters: &DBScanHyperParams,
+    observations: &ArrayBase<impl Data<Elem = f64>, Ix2>,
+) -> Array1<Option<usize>> {
+    let mut cluster_memberships = Array1::from_elem(observations.dim().1, None);
+    let mut current_cluster_id = 0;
+    for (i, obs) in observations.axis_iter(Axis(1)).enumerate() {
+        if cluster_memberships[i].is_some() {
+            continue;
+        }
+        let (neighbor_count, mut search_queue) = find_neighbors(
+            &obs,
+            observations,
+            hyperparameters.tolerance(),
+            &cluster_memberships,
+        );
+        if neighbor_count < hyperparameters.minimum_points() {
+            continue;
+        }
+        // Now go over the neighbours adding them to the cluster
+        cluster_memberships[i] = Some(current_cluster_id);
+
+        while !search_queue.is_empty() {
+            let candidate = search_queue.remove(0);
+
+            cluster_memberships[candidate.0] = Some(current_cluster_id);
+
+            let (_, mut neighbors) = find_neighbors(
+                &candidate.1,
+                observations,
+                hyperparameters.tolerance(),
+                &cluster_memberships,
+            );
+
+            search_queue.append(&mut neighbors);
+        }
+        current_cluster_id += 1;
+    }
+    cluster_memberships
+}
+
+fn find_neighbors<'a>(
+    candidate: &ArrayBase<impl Data<Elem = f64>, Ix1>,
+    observations: &'a ArrayBase<impl Data<Elem = f64>, Ix2>,
+    eps: f64,
+    clusters: &Array1<Option<usize>>,
+) -> (usize, Vec<(usize, ArrayView<'a, f64, Ix1>)>) {
+    let mut res = vec![];
+    let mut count = 0;
+    for (i, obs) in observations.axis_iter(Axis(1)).enumerate() {
+        if candidate.l2_dist(&obs).unwrap() < eps {
+            count += 1;
+            if clusters[[i]].is_none() {
```

Comment:
We are still using an index to access the cluster from `clusters`.
I expect a more significant performance improvement using `Zip`, so that we already have the value at hand.

Reply:
Ah I see. Tried it out and didn't see an improvement except with the small values:

```
dbscan/10               time:   [597.95 ns 599.53 ns 601.28 ns]                       
                        change: [-6.4851% -6.0215% -5.5629%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  6 (6.00%) high mild
  7 (7.00%) high severe
dbscan/100              time:   [3.8862 us 3.8933 us 3.9013 us]                        
                        change: [-0.2848% +0.1601% +0.6346%] (p = 0.51 > 0.05)
                        No change in performance detected.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe
dbscan/1000             time:   [36.933 us 37.005 us 37.095 us]                         
                        change: [-0.1611% +0.1888% +0.5086%] (p = 0.30 > 0.05)
                        No change in performance detected.
Found 8 outliers among 100 measurements (8.00%)
  5 (5.00%) high mild
  3 (3.00%) high severe
dbscan/10000            time:   [373.45 us 374.56 us 375.95 us]                         
                        change: [-0.5209% +0.1337% +0.9362%] (p = 0.73 > 0.05)
                        No change in performance detected.
Found 8 outliers among 100 measurements (8.00%)
  4 (4.00%) high mild
  4 (4.00%) high severe
```

Reply:
I used the `zip` method in the standard library `Iterator` trait not `ndarray::Zip` as I didn't expect there to be a different and the iterator one was easier to write quicker. I can try again with `ndarray::Zip` if there's any difference in implementation that could make it faster

![Screenshot from 2019-12-24 19-30-45](https://user-images.githubusercontent.com/3472518/71423829-ec954c00-2683-11ea-95a1-30ffef162d51.png)

Reply:
Just given it a shot - no significant differences.
The one thing we could do though is removing that `count` variable - accessing the length of a vector should be `O(1)` if I am not mistaken, so no need to keep track of it separately (in this way we can simply return a `Vec` from that function and use `.len()` to get `neighbors_count`).

Reply:
The count is there because count can be greater than the length of the vector. Remember how the list of neighbours was originally filtered to get the search queue because it could have points already allocated. The returned vector is the subset of neighbours that aren't yet allocated to a cluster

Reply:
Yes, you are right :+1: